File: clients/go/worker/jobPoller_test.go
// Copyright © 2018 Camunda Services GmbH (info@camunda.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package worker

import (
	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/suite"
	"github.com/zeebe-io/zeebe/clients/go/entities"
	"github.com/zeebe-io/zeebe/clients/go/mock_pb"
	"github.com/zeebe-io/zeebe/clients/go/pb"
	"github.com/zeebe-io/zeebe/clients/go/utils"
	"io"
	"math"
	"sync"
	"testing"
	"time"
)

type JobPollerSuite struct {
	suite.Suite
	ctrl      *gomock.Controller
	client    *mock_pb.MockGatewayClient
	poller    jobPoller
	waitGroup sync.WaitGroup
}

func (suite *JobPollerSuite) BeforeTest(suiteName, testName string) {
	suite.ctrl = gomock.NewController(suite.T())
	suite.client = mock_pb.NewMockGatewayClient(suite.ctrl)
	suite.poller = jobPoller{
		client:         suite.client,
		request:        pb.ActivateJobsRequest{},
		requestTimeout: utils.DefaultTestTimeout,
		maxJobsActive:  DefaultJobWorkerMaxJobActive,
		pollInterval:   DefaultJobWorkerPollInterval,
		jobQueue:       make(chan entities.Job),
		workerFinished: make(chan bool),
		closeSignal:    make(chan struct{}),
		remaining:      0,
		threshold:      int(math.Round(float64(DefaultJobWorkerMaxJobActive) * DefaultJobWorkerPollThreshold)),
	}
	suite.waitGroup.Add(1)
}

func (suite *JobPollerSuite) AfterTest(suiteName, testName string) {
	defer suite.ctrl.Finish()

	close(suite.poller.closeSignal)

	c := make(chan struct{})
	go func() {
		suite.waitGroup.Wait()
		close(c)
	}()

	for {
		select {
		case <-c:
			return
		case <-suite.poller.jobQueue:
			break // drop jobs
		case <-time.After(utils.DefaultTestTimeout):
			suite.FailNow("Failed to wait for job poller to close")
		}
	}
}

func TestJobPollerSuite(t *testing.T) {
	suite.Run(t, new(JobPollerSuite))
}

func (suite *JobPollerSuite) TestShouldPollAfterPollIntervalIfThresholdIsReached() {
	// given
	suite.poller.pollInterval = 250 * time.Millisecond
	suite.poller.maxJobsActive = 10
	suite.poller.threshold = 4

	gomock.InOrder(
		suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
			MaxJobsToActivate: 10,
		}}).Return(suite.singleJobStream(), nil),
		suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
			MaxJobsToActivate: 9,
		}}).Return(suite.singleJobStream(), nil),
	)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then two jobs are received (initial poll and after pollInterval)
	suite.consumeJob()
	suite.consumeJob()
}

func (suite *JobPollerSuite) TestShouldNotPollAfterIntervalIfNotThreshold() {
	// given
	suite.poller.pollInterval = 250 * time.Millisecond
	suite.poller.maxJobsActive = 10
	suite.poller.remaining = 6
	suite.poller.threshold = 4

	suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
		MaxJobsToActivate: 4,
	}}).Return(suite.singleJobStream(), nil)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then one job is received after inital poll and no further poll
	suite.consumeJob()
}

func (suite *JobPollerSuite) TestShoulPolldAfterJobFinsihedIfThresholdIsReached() {
	// given
	suite.poller.maxJobsActive = 10
	suite.poller.threshold = 4

	gomock.InOrder(
		suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
			MaxJobsToActivate: 10,
		}}).Return(suite.singleJobStream(), nil),
		suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
			MaxJobsToActivate: 10,
		}}).Return(suite.singleJobStream(), nil),
	)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then two jobs are received (initial poll and poll after job finished)
	suite.completeJob()
	suite.consumeJob()
}

func (suite *JobPollerSuite) TestShouldNotPollAfterJobIsFinishedIfNotThreshold() {
	// given
	suite.poller.maxJobsActive = 10
	suite.poller.remaining = 6
	suite.poller.threshold = 4

	suite.client.EXPECT().ActivateJobs(gomock.Any(), &utils.RpcTestMsg{Msg: &pb.ActivateJobsRequest{
		MaxJobsToActivate: 4,
	}}).Return(suite.singleJobStream(), nil)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then one job is received after inital poll and no further poll
	suite.completeJob()
}

func (suite *JobPollerSuite) TestShouldIgnoreRequestFailure() {
	// given
	suite.poller.maxJobsActive = 10
	suite.poller.remaining = 5
	suite.poller.threshold = 4

	gomock.InOrder(
		suite.client.EXPECT().ActivateJobs(gomock.Any(), gomock.Any()).Return(nil, io.ErrClosedPipe),
		suite.client.EXPECT().ActivateJobs(gomock.Any(), gomock.Any()).Return(suite.singleJobStream(), nil),
	)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then signal job finished to poll again
	suite.poller.workerFinished <- true

	// should receive another job
	suite.consumeJob()
}

func (suite *JobPollerSuite) TestShouldIgnoreStreamFailure() {
	// given
	suite.poller.maxJobsActive = 10
	suite.poller.remaining = 4
	suite.poller.threshold = 4

	gomock.InOrder(
		suite.client.EXPECT().ActivateJobs(gomock.Any(), gomock.Any()).Return(suite.streamFailureAfterFirstJob(), nil),
		suite.client.EXPECT().ActivateJobs(gomock.Any(), gomock.Any()).Return(suite.streamFailureAfterFirstJob(), nil),
	)

	// when
	go suite.poller.poll(&suite.waitGroup)

	// then signal job finished to poll again
	suite.completeJob()

	// should receive another job from broken stream
	suite.consumeJob()
}

func (suite *JobPollerSuite) singleJobStream() pb.Gateway_ActivateJobsClient {
	stream := mock_pb.NewMockGateway_ActivateJobsClient(suite.ctrl)
	gomock.InOrder(
		stream.EXPECT().Recv().Return(&pb.ActivateJobsResponse{
			Jobs: []*pb.ActivatedJob{
				{},
			},
		}, nil),
		stream.EXPECT().Recv().Return(nil, io.EOF),
	)

	return stream
}

func (suite *JobPollerSuite) streamFailureAfterFirstJob() pb.Gateway_ActivateJobsClient {
	stream := mock_pb.NewMockGateway_ActivateJobsClient(suite.ctrl)
	gomock.InOrder(
		stream.EXPECT().Recv().Return(&pb.ActivateJobsResponse{
			Jobs: []*pb.ActivatedJob{
				{},
			},
		}, nil),
		stream.EXPECT().Recv().Return(nil, io.ErrClosedPipe),
	)
	return stream
}

func (suite *JobPollerSuite) consumeJob() {
	<-suite.poller.jobQueue
}

func (suite *JobPollerSuite) completeJob() {
	suite.consumeJob()
	suite.poller.workerFinished <- true
}

File: clients/java/src/test/java/io/zeebe/client/job/JobPollerTest.java
/*
 * Copyright © 2017 camunda services GmbH (info@camunda.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.zeebe.client.job;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import io.zeebe.client.api.response.ActivatedJob;
import io.zeebe.client.impl.ZeebeObjectMapper;
import io.zeebe.client.impl.worker.JobPoller;
import io.zeebe.client.util.ClientTest;
import io.zeebe.gateway.protocol.GatewayOuterClass;
import io.zeebe.gateway.protocol.GatewayOuterClass.ActivateJobsRequest;
import java.time.Duration;
import java.util.function.Consumer;
import java.util.function.IntConsumer;
import org.awaitility.Awaitility;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

public final class JobPollerTest extends ClientTest {

  private Consumer<ActivatedJob> jobConsumer;
  private IntConsumer doneCallback;
  private Consumer<Throwable> errorCallback;

  @Before
  public void setup() {
    jobConsumer = Mockito.spy(Consumer.class);
    doneCallback = Mockito.spy(IntConsumer.class);
    errorCallback = Mockito.spy(Consumer.class);
  }

  @Test
  public void shouldSetRequestTimeout() {
    // given
    final Duration requestTimeout = Duration.ofHours(123);
    final JobPoller jobPoller = getJobPoller(requestTimeout);

    // when
    jobPoller.poll(123, jobConsumer, doneCallback, errorCallback, () -> true);

    // then
    rule.verifyRequestTimeout(requestTimeout);
  }

  @Test
  public void shouldCallbackWhenPollComplete() {
    // given
    gatewayService.onActivateJobsRequest(TestData.job(), TestData.job());

    // when
    getJobPoller().poll(123, jobConsumer, doneCallback, errorCallback, () -> true);

    // then
    Awaitility.await()
        .atMost(Duration.ofSeconds(5))
        .untilAsserted(
            () -> {
              verify(jobConsumer, times(2)).accept(any(ActivatedJob.class));
              verify(doneCallback).accept(eq(2));
              verify(errorCallback, never()).accept(any(Throwable.class));
            });
  }

  @Test
  public void shouldCallbackWhenPollFailed() {
    // given
    gatewayService.onActivateJobsRequest(new StatusRuntimeException(Status.RESOURCE_EXHAUSTED));

    // when
    getJobPoller().poll(123, jobConsumer, doneCallback, errorCallback, () -> true);

    // then
    Awaitility.await()
        .atMost(Duration.ofSeconds(5))
        .untilAsserted(
            () -> {
              verify(jobConsumer, never()).accept(any(ActivatedJob.class));
              verify(doneCallback, never()).accept(any(Integer.class));
              verify(errorCallback).accept(any(StatusRuntimeException.class));
            });
  }

  private JobPoller getJobPoller() {
    return getJobPoller(Duration.ofSeconds(10));
  }

  private JobPoller getJobPoller(final Duration requestTimeout) {
    return new JobPoller(
        rule.getGatewayStub(),
        ActivateJobsRequest.newBuilder(),
        new ZeebeObjectMapper(),
        requestTimeout,
        (t) -> false);
  }

  private static final class TestData {
    private static GatewayOuterClass.ActivatedJob job() {
      return GatewayOuterClass.ActivatedJob.newBuilder()
          .setKey(12)
          .setType("foo")
          .setWorkflowInstanceKey(123)
          .setBpmnProcessId("test1")
          .setWorkflowDefinitionVersion(2)
          .setWorkflowKey(23)
          .setElementId("foo")
          .setElementInstanceKey(23213)
          .setCustomHeaders("{\"version\": \"1\"}")
          .setWorker("worker1")
          .setRetries(34)
          .setDeadline(1231)
          .setVariables("{\"key\": \"val\"}")
          .build();
    }
  }
}
